### PR TITLE
WIP: mock up of new API

### PR DESCRIPTION
**Pull request by jasmainak**
alternative to #106 after discussion with @agramfort

### CODE REVIEW

File: mne_bids/mne_bids.py [hunk 1]
```diff
@@ -528,22 +528,21 @@ def _sidecar_json(raw, task, manufacturer, fname, kind, eeg_ref=None,
...
 
-def raw_to_bids(subject_id, task, raw_file, output_path, session_id=None,
-                acquisition=None, run=None, kind='meg', events_data=None,
-                event_id=None, hpi=None, electrode=None, hsp=None,
-                eeg_ref=None, eeg_gnd=None, config=None,
-                overwrite=False, verbose=True):
+def convert_to_bids(raw_fname, output_path, subject_id, task, session_id=None,
+                    acquisition=None, run=None, processing=None, kind='auto',
+                    events_data=None, event_id=None, overwrite=False,
+                    verbose=True, **kwargs):
```
Comment (jasmainak) [inline mne_bids/mne_bids.py]: @agramfort @teonbrooks take a look at this new API

Just some thoughts that pop up in my head as I think about this:

- while we do not expose the quirks of the MNE I/O API anymore, the user still will have to use a dozen or so arguments when _calling_ this function. It's the cost you pay for having everything in one function.

- do we eventually also implement a `write_raw_bids` function or would you consider that off the table? 
    - Pros
         - Useful if the user wants to edit/correct some information before writing the file
         - We're not using hacks (c.f., `raw._orig_units`) to get around the issue of writing to certain formats like `edf`.
         - Promotes writing to a single file format and standardization
    - Cons
         - "two ways of doing the same thing" (same thing being writing BIDS compatible folders in `fif` format)
         -  `fif` format is not supported by BIDS-EEG and BIDS-iEEG
         - more work to implement
         - would limit adoption + user base of `mne-bids` as many people (cc @sappelhoff @monkeyman192) may want to retain original format

all this is to just make sure you've thought through the different pros and cons. For me, it's easiest to implement the kwargs alternative :-)

Comment (jasmainak) [reply mne_bids/mne_bids.py]: Just to illustrate, one function call from test currently looks like this:

```py
    raw_to_bids(subject_id=subject_id, session_id=session_id, run=run,
               task=task, acquisition=acq, raw_file=raw_fname,
               config=config_fname, hsp=headshape_fname,
               output_path=output_path)
```

This is not going to change even if we use `kwargs` ...

**Comment by agramfort** [comment]
comments :

- I would rename kwargs to raw_params so it's explicit it will be passed to read_raw_xxx
- I would still call it raw_to_bids but raw should be the first argument
- if raw is a string you write in the original format
- if raw is an instance of BaseRaw then you write to fif as you don't know what was changed in the raw instance.
- one option would be to be able to pass a new info so API could be:

```
raw_to_bids(raw, subject_id, task, output_path, session_id=None,
                     acquisition=None, run=None, kind='meg', events=None,
                     event_id= None, annotations=None, info=None,
                     **raw_params)
```

**Comment by teonbrooks** [comment]
>     * I would rename kwargs to raw_params so it's explicit it will be passed to read_raw_xxx

this is similar to the proposal of @monkeyman192 https://github.com/mne-tools/mne-bids/issues/98#issuecomment-424926555.

I think this would be correct way to go.

@agramfort , wrt events and annotations, are we handling annotation files or would this be the Annotation object or would this be event-like annotations from experimenter?

**Comment by agramfort** [comment]
regarding annotations it would be an MNE Annotations object. Not file on
disk.

**Comment by agramfort** [comment]
what do you think of my proposal to be able to pass an info param?

**Comment by jasmainak** [comment]
oops, sorry I didn't notice that. This is closer to what I had in mind. However, I would ask what you would do in the following scenarios:

```py
raw_to_bids(raw, ..., info)
```
do you use `raw.info` or `info` ? Also, I can think of potentially unsafe scenarios here too, like:

```
raw = mne.io.read_raw_fif(raw_fname, preload=True)
raw.pick_types(meg='grad')
info = raw.info.copy()
raw_to_bids(raw_fname, ..., info)
```

**Comment by monkeyman192** [comment]
> Also, I can think of potentially unsafe scenarios here too, like:

Here though the user is going out of their way to mess it up. I feel like we shouldn't have to account for something like that...

Also, I agree that passing in the raw object does save a lot of hassle in terms of passing in parameters. For example if we add the `.hsp`, .electrode` and `.hpi` attributes to a raw object then we automatically remove 3 arguments for MNE-BIDS which is obviously a good thing.

wrt. `info`, I feel this should just be used for data that cannot be stored in the info of the MNE raw objects. For example the group the participant is in or the project name (although project name should probably be an actual argument since I feel it is *kinda* important...

**Comment by agramfort** [comment]
>
> Also, I can think of potentially unsafe scenarios here too, like:
>
>
yes you would need safeguards to check that the number of channels is the
same.

if you pass a raw instance really I don't know what to expect from users.

and if you have  a parameter called info please make sure that is is a
standard
mne Info instance.

What do you think is the issue of doing:

write_raw_bids('test_pdf_linux', output_path=output_path,
subject_id=subject_id, session_id=session_id, run=run,
                task=task, acquisition=acq, config=config_fname,
hsp=headshape_fname)

where config=config_fname, hsp=headshape_fname are **raw_params

**Comment by jasmainak** [comment]
IMO, you would have potentially the same issues with an `info` instance that you would have with `raw` that is not preloaded (except maybe the cropping issue that you mentioned which is a problem only with `raw` instance) + you are making two accesses to the disk (once for modifying the `info` and once inside `raw_to_bids`/`write_raw_bids`)

So, I guess it's more a question of where you want to go with the API. If you want to tell people: "hey, we let you convert a raw file into BIDS -- it's a black box how we do it but it does the job for you." vs. "let's read in a raw object, add the info we need and write it out. And once we have the reading functionality, you can read the BIDS data, modify the raw/info and write it back (instead of having to edit the files manually or writing your own functions for doing it)." The second option is more versatile I feel but I can see that it can sometimes be not as safe. 

The current API is basically historical cruft. It's because @teonbrooks and I wanted to push for adoption of a single file format (`fif`), we wrote a function that can take in any of the params of `read_raw_xxx` and write out a `fif`. However, `mne-bids` now also does EEG (and `fif` is not supported there) and it retains the original manufacturer format (by copying the files from the original folder). 

Also, I dislike that you have to use so many parameters to call the function but I defer to you for judgements in Pythonic taste :)

**Pull request closed without merging**